### PR TITLE
Fsa py rewrite

### PR DESCRIPTION
**Pull request by danpovey**

### CODE REVIEW

**Comment by danpovey** [comment]
Rewriting fsa.py, simplify some things a little, make others easier to understand.

**Comment by danpovey** [comment]
Note: we need some small change to downstream code; fsa.to_(device) no longer exists, you need: fsa = fsa.to(device)

File: k2/python/k2/autograd_utils.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        assert torch.all(torch.eq(out_fsa.scores, unused_in_fsa_scores))
+        return out_fsa.scores
+
+    @staticmethod
+    def backward(ctx, out_fsa_scores_grad: torch.Tensor) -> Tuple[None, torch.Tensor]:
+        return None, out_fsa_scores_grad
+
+def phantom_set_scores_to(fsa, scores_value) -> None:
+    # we don't need the output value of the following call
+    # (which it fsa.score), since it is accessible through `fsa`.
```
Comment (csukuangfj) [inline k2/python/k2/autograd_utils.py:32]: typo:

which is fsa.scores

File: k2/python/k2/fsa.py [hunk 2]
```diff
@@ -227,92 +251,156 @@ def __setattr__(self, name: str, value: Any) -> None:
...
+        if properties & fsa_properties.VALID != 1:
+            raise ValueError(
+                "Fsa is not valid, properties are: {} = {}, arcs are: {}".
+                format(properties, fsa_properties.to_str(properties),
+                       str(self.arcs)))
+        return properties
+
+    @properties.setter
+    def properties(self, value) -> None:
+        """ Only supports setting self.property to None; any other
```
Comment (csukuangfj) [inline k2/python/k2/fsa.py:301]: ~~self.property~~,  self.properties

File: k2/python/k2/fsa.py [hunk 3]
```diff
@@ -227,92 +251,156 @@ def __setattr__(self, name: str, value: Any) -> None:
...
+        value should be set internally by writing to self.__dict__ directly."""
+        if value == None:
+            self.__dict__['_properties'] = None
+        else:
+            raise RuntimeError("""Currently we don't allow the .properties of an Fsa
+            to be set this way, except to None""");
+
+    @property
+    def properties_str(self) -> str:
+        return _k2.fsa_properties_as_str(self._properties)
```
Comment (csukuangfj) [inline k2/python/k2/fsa.py:311]: Should it be `self.properties` here?

`self._properties` may be None.

File: k2/python/k2/fsa.py [hunk 4]
```diff
@@ -227,92 +251,156 @@ def __setattr__(self, name: str, value: Any) -> None:
...
             del self._non_tensor_attr[name]
-        elif name in self._grad_cache:
-            del self._grad_cache[name]
-        else:
-            super().__delattr__(name)
-
-    def _update_cache(self, name: str, value: Any) -> None:
-        self._grad_cache[name] = value
+        elif name in self._cache:
+            del self._cache[name]
```
Comment (csukuangfj) [inline k2/python/k2/fsa.py:357]: What if all the above tests fail?

Should we raise an exception for deleting a non-existent attribute?

File: k2/python/k2/fsa.py [hunk 5]
```diff
@@ -227,92 +251,156 @@ def __setattr__(self, name: str, value: Any) -> None:
...
+    @property
+    def properties_str(self) -> str:
+        return _k2.fsa_properties_as_str(self._properties)
+
+    @property
+    def requires_grad(self) -> bool:
+        return self.scores.requires_grad
+
+    @property
+    def grad(self) -> bool:
```
Comment (qindazhu) [inline k2/python/k2/fsa.py:318]: why is it a bool?

Comment (csukuangfj) [reply k2/python/k2/fsa.py:318]: I think this must be a  typo introduced by copying and pasting `def requires_grad`.

File: k2/python/k2/fsa.py [hunk 6]
```diff
@@ -599,43 +685,107 @@ def from_dict(cls, dict_in: Dict[str, Any]) -> 'Fsa':
...
-    def named_tensor_attr(self) -> Iterator[Tuple[str, torch.Tensor]]:
+        return ans
+
+    def detach(self) -> 'Fsa':
+        """
+        Return an Fsa that shares the underlying data with this one,
+        except gradients are not tracked.
+        Any non-tensor attributes are copied over.
+        """
+        ans = Fsa(self.arcs.detach(), properties=self.properties)
```
Comment (csukuangfj) [inline k2/python/k2/fsa.py:756]: So where is the method `detach` of `arcs` defined?

File: k2/python/k2/fsa.py [hunk 7]
```diff
@@ -770,13 +889,27 @@ def from_openfst(cls, s: str, acceptor: bool = True) -> 'Fsa':
...
+# value but we know they *already have that value*.
+#
+# The 'forward' function "pretends" to set out_fsa.scores to
+# 'unused_in_fsa_scores', and it returns out_fsa.scores.  This will
+# attach a _grad_fn to out_fsa.scores, if unused_in_fsa_scores.requires_grad
+# was true.
+#
+# The backprop is as if the function was just a copy (i.e. it copies
+# the output gradient)
+class _PhantomSetScoresFunction(torch.autograd.Function):
```
Comment (qindazhu) [inline k2/python/k2/fsa.py:907]: seems duplicate code with autograd.py?

File: k2/python/k2/fsa.py (hunk 3 above)
Comment (danpovey) [reply k2/python/k2/fsa.py:311]: Changing it.  Usually self._properties will be there (it's defined in self.__dict__) but safer to access self.properties

File: k2/python/k2/fsa.py (hunk 4 above)
Comment (danpovey) [reply k2/python/k2/fsa.py:357]: OK will raixe

File: k2/python/k2/fsa.py (hunk 6 above)
Comment (danpovey) [reply k2/python/k2/fsa.py:756]: Ah yes, detach() is not needed on arcs, it's not a tensor.

File: k2/python/k2/fsa.py (hunk 7 above)
Comment (danpovey) [reply k2/python/k2/fsa.py:907]: yes, removing

**Comment by danpovey** [comment]
You will have seen I didn't test everything, e.g. detach()-- sorry.
If anyone has time to write test code it would be great.
